geoips/utils/types/datatree_ditto.py

```diff
@@ -0,0 +1,600 @@
+            DataTreeDitto._converters = {}
+            DataTreeDitto._register_builtin_converters()
+
+        # Convert data if the data is not an xarray DataSet
+        if dataset is not None:
+            # if isinstance(dataset, xr.DataArray):
+            #    # Convert DataArray to Dataset
+            #    dataset = dataset.to_dataset()
+            if not isinstance(
+                dataset,
+                (
+                    xr.Dataset,
+                    xr.DataArray,
+                    xr.core.coordinates.DataTreeCoordinates,
+                    xr.DataTree,
+                ),
+            ):
+                # Convert non-xarray object
+                dataset = self._convert_to_dataset(dataset)
+
+        super().__init__(dataset=dataset, children=children, name=name)
+
+    def _convert_output_to_DataTreeDitto(f):
+        def f_that_returns_datatree_ditto(*args, **kwargs):
+            result = f(*args, **kwargs)
+            if isinstance(result, DataTree) and not isinstance(result, DataTreeDitto):
+                return DataTreeDitto._convert_datatree_to_ditto(result)
+            return result
+
+        return f_that_returns_datatree_ditto
+
+    @classmethod
+    def _register_builtin_converters(cls):
+        """Register built-in converters for common types.
+
+        Registers converters for numpy arrays automatically when the class
+        is first instantiated.
+        """
+        cls.register_converter(np.ndarray, cls._numpy_to_dataset, cls._dataset_to_numpy)
+
+    @classmethod
+    def register_converter(
+        cls,
+        obj_type: type,
+        to_dataset_func: Callable,
+        from_dataset_func: Callable,
+    ) -> None:
+        """Register a converter for a specific object type.
+
+        Parameters
+        ----------
+        obj_type : type
+            The type of object to convert.
+        to_dataset_func : callable
+            Function to convert object to xarray.Dataset. Must accept the object
+            as first argument and return an xarray.Dataset.
+        from_dataset_func : callable
+            Function to convert xarray.Dataset back to original object. Must
+            accept an xarray.Dataset and return an object of obj_type.
+
+        Examples
+        --------
+        >>> def list_to_dataset(lst, name="data", **kwargs):
+        ...     arr = np.array(lst)
+        ...     return DataTreeDitto._numpy_to_dataset(arr, name, **kwargs)
+        >>> def dataset_to_list(ds, **kwargs):
+        ...     return DataTreeDitto._dataset_to_numpy(ds, **kwargs).tolist()
+        >>> DataTreeDitto.register_converter(list, list_to_dataset, dataset_to_list)
+        """
+        # if not hasattr(DataTreeDitto, "_converters"):
+        #    DataTreeDitto._converters = {}
+        #    DataTreeDitto._register_builtin_converters()
+        cls._converters[obj_type] = {
+            "to_dataset": to_dataset_func,
+            "from_dataset": from_dataset_func,
+        }
```

Comment:
Given that these are registered by their data type, will it correctly handle data types that are subclasses of registered data types?

I think this is actually two questions:
1. If both `numpy.array` and `numpy.masked_array` are registered, would DataTreeDitto always select the correct type?
2. If only `numpy.array` is registered, would DataTreeDitto attempt to handle `numpy.masked_array` since it is a subclass of `numpy.array` or would it error because `numpy.masked_array` isn't registered?

Reply:
1. No, this is a current limitation of this approach using duck-typing.
2. Yes

As we talked about, it is likely that we could solve this by: type checking more aggressively OR creating converters that handle types and their children. In this example, it would be a converter that handles numpy arrays and masked arrays.